File: backend/app/schemas/donation.py
"""Pydantic schemas for Surplus Food Donations."""

from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field, model_validator
from app.schemas.need import DietType


class StorageCondition(str, Enum):
    AMBIENT = "ambient"
    REFRIGERATED = "refrigerated"
    FROZEN = "frozen"
    HOT_HELD = "hot-held"


class DonationStatus(str, Enum):
    DRAFT = "DRAFT"
    POSTED = "POSTED"
    MATCHING = "MATCHING"
    MATCHED = "MATCHED"
    ALLOCATED = "ALLOCATED"
    CANCELLED = "CANCELLED"
    EXPIRED = "EXPIRED"


class DonationBase(BaseModel):
    diet_type: DietType
    food_category_id: str = Field(..., description="Food Category UUID")
    declared_quantity_kg: float = Field(..., ge=5.0, description="Declared quantity in kg (minimum 5 kg)")
    preparation_at: str = Field(..., description="ISO 8601 preparation timestamp")
    available_from: str = Field(..., description="ISO 8601 pickup availability start")
    rescue_deadline: str = Field(..., description="ISO 8601 safe consumption/rescue deadline")
    storage_condition: StorageCondition = StorageCondition.AMBIENT
    packaging_type: str = Field(..., min_length=2, max_length=150)
    allergen_information: Optional[str] = None
    raw_description: str = Field(..., min_length=3, max_length=1000)
    normalized_description: Optional[str] = None
    image_path: Optional[str] = None
    pickup_location_id: str = Field(..., description="Donor pickup location UUID")

    @model_validator(mode="after")
    def validate_deadline_after_availability(self) -> "DonationBase":
        if self.rescue_deadline <= self.available_from:
            raise ValueError("rescue_deadline must be strictly after available_from.")
        return self


class DonationCreate(DonationBase):
    pass


class DonationUpdate(BaseModel):
    diet_type: Optional[DietType] = None
    food_category_id: Optional[str] = None
    declared_quantity_kg: Optional[float] = Field(None, ge=5.0)
    preparation_at: Optional[str] = None
    available_from: Optional[str] = None
    rescue_deadline: Optional[str] = None
    storage_condition: Optional[StorageCondition] = None
    packaging_type: Optional[str] = None
    allergen_information: Optional[str] = None
    raw_description: Optional[str] = None
    normalized_description: Optional[str] = None
    image_path: Optional[str] = None
    pickup_location_id: Optional[str] = None


class DonationResponse(BaseModel):
    id: str
    donor_id: str
    diet_type: DietType
    food_category_id: str
    declared_quantity_kg: float
    remaining_quantity_kg: float
    preparation_at: str
    available_from: str
    rescue_deadline: str
    storage_condition: StorageCondition
    packaging_type: str
    allergen_information: Optional[str] = None
    raw_description: str
    normalized_description: Optional[str] = None
    image_path: Optional[str] = None
    pickup_location_id: str
    status: DonationStatus
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

File: anna-setu-frontend-development/lib/api/donation.ts
/**
 * AnnaSetu — Surplus Food Donation API Client
 */

import { apiClient } from './client'
import type { DietType } from './need'

export type StorageCondition = 'ambient' | 'refrigerated' | 'frozen' | 'hot-held'
export type DonationStatus = 'DRAFT' | 'POSTED' | 'MATCHING' | 'MATCHED' | 'ALLOCATED' | 'CANCELLED' | 'EXPIRED'

export interface DonationResponse {
  id: string
  donor_id: string
  diet_type: DietType
  food_category_id: string
  declared_quantity_kg: number
  remaining_quantity_kg: number
  preparation_at: string
  available_from: string
  rescue_deadline: string
  storage_condition: StorageCondition
  packaging_type: string
  allergen_information?: string
  raw_description: string
  normalized_description?: string
  image_path?: string
  pickup_location_id: string
  status: DonationStatus
  visual_quality_assessment?: FoodQualityAssessment
  created_at?: string
  updated_at?: string
}

export interface FoodQualityAssessment {
  donation_id?: string
  food_identified?: string
  visual_quality_score: number
  freshness_signal: 'GOOD' | 'FAIR' | 'CONCERN' | 'UNKNOWN'
  packaging_condition: 'GOOD' | 'FAIR' | 'POOR' | 'NOT_VISIBLE'
  image_quality: 'GOOD' | 'FAIR' | 'INSUFFICIENT'
  visible_concerns: string[]
  risk_flags: string[]
  confidence: number
  recommendation: 'VISUAL_REVIEW_PASS' | 'MANUAL_REVIEW' | 'INSUFFICIENT_IMAGE'
  explanation: string
  disclaimer: string
  is_cached?: boolean
  provider?: string
  analyzed_at?: string
}

export interface FoodQualityCheckPayload {
  image_url?: string
  image_base64?: string
}

export interface DonationCreatePayload {
  diet_type: DietType
  food_category_id: string
  declared_quantity_kg: number
  preparation_at: string
  available_from: string
  rescue_deadline: string
  storage_condition?: StorageCondition
  packaging_type: string
  allergen_information?: string
  raw_description: string
  normalized_description?: string
  image_path?: string
  pickup_location_id: string
}

export interface DonationUpdatePayload {
  diet_type?: DietType
  food_category_id?: string
  declared_quantity_kg?: number
  preparation_at?: string
  available_from?: string
  rescue_deadline?: string
  storage_condition?: StorageCondition
  packaging_type?: string
  allergen_information?: string
  raw_description?: string
  normalized_description?: string
  image_path?: string
  pickup_location_id?: string
}

export const donationApi = {
  /** Create a new surplus food donation in DRAFT status (min 5 kg) */
  async createDonation(payload: DonationCreatePayload): Promise<DonationResponse> {
    return apiClient<DonationResponse>('/api/v1/donations', {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },

  /** List donations owned by the donor */
  async listDonations(status?: DonationStatus): Promise<DonationResponse[]> {
    const query = status ? `?status_filter=${status}` : ''
    return apiClient<DonationResponse[]>(`/api/v1/donations${query}`)
  },

  /** Get specific donation details */
  async getDonation(donationId: string): Promise<DonationResponse> {
    return apiClient<DonationResponse>(`/api/v1/donations/${donationId}`)
  },

  /** Update editable donation details */
  async updateDonation(donationId: string, payload: DonationUpdatePayload): Promise<DonationResponse> {
    return apiClient<DonationResponse>(`/api/v1/donations/${donationId}`, {
      method: 'PATCH',
      body: JSON.stringify(payload),
    })
  },

  /** Post surplus food donation for matching */
  async postDonation(donationId: string): Promise<DonationResponse> {
    return apiClient<DonationResponse>(`/api/v1/donations/${donationId}/post`, {
      method: 'POST',
    })
  },

  /** Cancel donation */
  async cancelDonation(donationId: string): Promise<DonationResponse> {
    return apiClient<DonationResponse>(`/api/v1/donations/${donationId}/cancel`, {
      method: 'POST',
    })
  },

  /** Run AI visual quality check on donation image */
  async checkFoodQuality(donationId: string, payload?: FoodQualityCheckPayload): Promise<FoodQualityAssessment> {
    return apiClient<FoodQualityAssessment>(`/api/v1/donations/${donationId}/ai-quality-check`, {
      method: 'POST',
      body: payload ? JSON.stringify(payload) : undefined,
    })
  },
}
